### PR TITLE
CORCI-1010 test: Change to running NLT on real, not virtual machines.

### PR DESCRIPTION
**Pull request by ashleypittman**
Add a NLT stage to unit test to run just the node local testing.
Add Skip-unit-tests option to skip all unit testing, and Skip-nlt to skip this new stage.
Change NLT to run on real hardware rather than a VM to reduce runtime and allow
larger tests to be added later.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -1052,15 +1053,35 @@ pipeline {
...
                     }
                     steps {
-                        unitTest timeout_time: 60,
+                        unitTest timeout_time: 30,
+                                 inst_repos: pr_repos(),
+                                 inst_rpms: unit_packages()
+                    }
+                    post {
+                      always {
+                            unitTestPost artifacts: ['unit_test_logs/*'],
```
Comment (brianjmurrell) [inline Jenkinsfile:1062]:
```suggestion
                          unitTestPost artifacts: ['unit_test_logs/*'],
```

File: ci/unit/test_main_node.sh [hunk 2]
```diff
@@ -44,15 +44,18 @@ else
 fi
 
 cd "$DAOS_BASE"
-IS_CI=true OLD_CI=false RUN_TEST_VALGRIND="$WITH_VALGRIND" utils/run_test.sh
-mkdir -p vm_test
+if [ -z "$NLT" ]; then
```
Comment (brianjmurrell) [inline ci/unit/test_main_node.sh:47]: Since `$NLT` is boolean:

```suggestion
if ${NLT:-false}; then
```

File: ci/unit/test_main_node.sh [hunk 3]
```diff
@@ -44,15 +44,18 @@ else
...
 
 # Remove DAOS_BASE from memcheck xml results
 set -x
 if [ "$WITH_VALGRIND" == 'memcheck' ]; then
     find test_results -maxdepth 1 -name '*.memcheck.xml' \
         -print0 | xargs -0 sed -i "s:$DAOS_BASE::g"
     mv test_results/unit-test-*.memcheck.xml .
-elif [ -z "$BULLSEYE" ]; then
+fi
+if [ -n "$NLT" ]; then
```
Comment (brianjmurrell) [inline ci/unit/test_main_node.sh:58]: Ditto:
```suggestion
if ${NLT:-false}; then
```

File: ci/unit/test_main.sh [hunk 4]
```diff
@@ -27,6 +27,11 @@ else
   BULLSEYE=
 fi
 
+NLT=
```
Comment (brianjmurrell) [inline ci/unit/test_main.sh:30]: Booleans should always have a `true`/`false` value.

```suggestion
NLT=false
```

File: ci/unit/test_main_node.sh [hunk 5]
```diff
@@ -44,15 +44,18 @@ else
...
 
 cd "$DAOS_BASE"
-IS_CI=true OLD_CI=false RUN_TEST_VALGRIND="$WITH_VALGRIND" utils/run_test.sh
-mkdir -p vm_test
+if [ -z "$NLT" ]; then
+    IS_CI=true OLD_CI=false RUN_TEST_VALGRIND="$WITH_VALGRIND" utils/run_test.sh
+fi
 
 # Remove DAOS_BASE from memcheck xml results
 set -x
```
Comment (brianjmurrell) [inline ci/unit/test_main_node.sh:52]: This is superfluous, can you remove it in this PR please?

**Review by brianjmurrell — CHANGES_REQUESTED** [review]
It seems like https://github.com/daos-stack/daos/blob/9a0c49909fe78941cacfd666cf9123ddc8702559/ci/unit/test_main_node.sh#L47-L61 are really two different test paths for `run_test.sh` and `node_local_test.py`, where the lines before those two paths are just common pre-test setup.

I wonder if it's worth splitting that file into two different tests that each source the common setup.

Have these changes been tested with Bullseye enabled?

File: Jenkinsfile (hunk 1 above)
Comment (ashleypittman) [reply Jenkinsfile:1062]: I don't quite get this change, but there were some whitespace issues on the following line which I've resolved.

**Comment by ashleypittman** [comment]
> It seems like
> 
> https://github.com/daos-stack/daos/blob/9a0c49909fe78941cacfd666cf9123ddc8702559/ci/unit/test_main_node.sh#L47-L61
> 
> are really two different test paths for `run_test.sh` and `node_local_test.py`, where the lines before those two paths are just common pre-test setup.
> I wonder if it's worth splitting that file into two different tests that each source the common setup.

I did think about this, but you'd need to duplicate test_main.sh, test_main_node.sh, test_post_always.sh, test_post_always_node.sh, as well as unitTestPost.groovy, and as you say the bulk of the logic is in the setup which is actually common anyway.

> Have these changes been tested with Bullseye enabled?

No, I did try to walk through the logic and couldn't work out how Bullseye worked anyway.  I'll try it now.

**Comment by ashleypittman** [comment]
Bullseye is failing even without this PR.

**Comment by ashleypittman** [comment]
This is ready for review now, it needs https://github.com/daos-stack/pipeline-lib/pull/167 to land first, and https://github.com/daos-stack/daos/pull/3792 (whilst it doesn't want to be landed at all) verifies that the pipeline-lib changes do not break current daos.

**Review by JohnMalmberg — APPROVED** [review]
We have a lot more VMs than we have physical machines, so a job may end up waiting longer for a physical machine that it would take for a run on a VM.

**Comment by ashleypittman** [comment]
> We have a lot more VMs than we have physical machines, so a job may end up waiting longer for a physical machine that it would take for a run on a VM.

Yes, that is a concern and we'll have to see how that goes.  I've got some other PRs in flight that add microbenchmarks to this stage, but if run on VMs they're both slow and also give inconsistent results so real hardware is much preferable.

Even if we do have to back this that part of this PR, and make it run on VMs again then it's still a step forward having this though, as it makes the stages clearer and independent of each other.

**Review by brianjmurrell — DISMISSED** [review]
> I did think about this, but you'd need to duplicate test_main.sh,

Yes.  But the actual duplication would be limited to sourcing the setup.  The rest would be different.

> test_main_node.sh, test_post_always.sh, test_post_always_node.sh, as well as unitTestPost.groovy,

Would those need duplicating?  They would just be common to the two stages wouldn't they?

**Comment by ashleypittman** [comment]
> > I did think about this, but you'd need to duplicate test_main.sh,
> 
> Yes. But the actual duplication would be limited to sourcing the setup. The rest would be different.
> 
> > test_main_node.sh, test_post_always.sh, test_post_always_node.sh, as well as unitTestPost.groovy,
> 
> Would those need duplicating? They would just be common to the two stages wouldn't they?

I'll experiment with this in the next PR.

**Comment by ashleypittman** [comment]
Same code changes, just rebased and with the change to use the NLT branch of pipeline-lib removed now that's landed.  Please re-review so this can also land.

**Review by brianjmurrell — APPROVED** [review]

**Review by JohnMalmberg — APPROVED** [review]

**Comment by ashleypittman** [comment]
Failures are DAOS-5992, DAOS-6026, possibly DAOS-6954 (evict test timing out, unrelated to this PR), and "DAOS vc tests.VC7: verify with lost dkey" test which is also failing on master but doesn't appear to have a ticket.

**Comment by brianjmurrell** [comment]
> "DAOS vc tests.VC7: verify with lost dkey" test which is also failing on master but doesn't appear to have a ticket.

FWIW, the process is that if one cannot find a ticket, one should open the new ticket and reference it in the PR with an indication that it's a new ticket/failure.

**Pull request merged**